Asset loading test runner

ycombinator opened the pull request:
This PR implements a new type of test runner: `asset`. This test runner tests whether a package's assets (ES index templates, Kibana dashboards, etc.) are loaded up as expected.

## Sample usage

```
$ elastic-package test asset
Run asset tests for the package
--- Test results for package: apache - START ---
╭─────────┬─────────────┬───────────┬────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                                                      │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ apache  │             │ asset     │ dashboard apache-Logs-Apache-Dashboard is loaded               │ PASS   │ 6.101128287s │
│ apache  │             │ asset     │ dashboard apache-Metrics-Apache-HTTPD-server-status is loaded  │ PASS   │        927ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-CPU is loaded                       │ PASS   │        620ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Hostname-list is loaded             │ PASS   │        512ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Load1-slash-5-slash-15 is loaded    │ PASS   │        526ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Scoreboard is loaded                │ PASS   │        430ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Total-accesses-and-kbytes is loaded │ PASS   │        424ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Uptime is loaded                    │ PASS   │        416ns │
│ apache  │             │ asset     │ visualization apache-HTTPD-Workers is loaded                   │ PASS   │        403ns │
│ apache  │             │ asset     │ visualization apache-access-unique-IPs-map is loaded           │ PASS   │        404ns │
│ apache  │             │ asset     │ visualization apache-browsers is loaded                        │ PASS   │        413ns │
│ apache  │             │ asset     │ visualization apache-error-logs-over-time is loaded            │ PASS   │        410ns │
│ apache  │             │ asset     │ visualization apache-operating-systems is loaded               │ PASS   │        407ns │
│ apache  │             │ asset     │ visualization apache-response-codes-of-top-URLs is loaded      │ PASS   │        439ns │
│ apache  │             │ asset     │ visualization apache-response-codes-over-time is loaded        │ PASS   │        453ns │
│ apache  │             │ asset     │ search apache-HTTPD is loaded                                  │ PASS   │        598ns │
│ apache  │             │ asset     │ search apache-access-logs is loaded                            │ PASS   │        420ns │
│ apache  │             │ asset     │ search apache-errors-log is loaded                             │ PASS   │        514ns │
│ apache  │ access      │ asset     │ index_template logs-apache.access is loaded                    │ PASS   │        490ns │
│ apache  │ access      │ asset     │ ingest_pipeline logs-apache.access-0.2.7 is loaded             │ PASS   │        462ns │
│ apache  │ error       │ asset     │ index_template logs-apache.error is loaded                     │ PASS   │        405ns │
│ apache  │ error       │ asset     │ ingest_pipeline logs-apache.error-0.2.7 is loaded              │ PASS   │        434ns │
│ apache  │ status      │ asset     │ index_template metrics-apache.status is loaded                 │ PASS   │        440ns │
╰─────────┴─────────────┴───────────┴────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: apache - END   ---
Done
```

### TODO
* [x] Install package
* [x] Remove package
* [x] Add assertions on asset existence
* [x] Move `--data-stream` global test flag to pipeline and system tests
* [x] Add HOWTO doc
* [x] Add to README doc
* [x] Update CI configuration

Resolves #115.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-168/detail/PR-168/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-168/detail/PR-168/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-168/detail/PR-168/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-168/detail/PR-168/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_168.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #168 opened]`

* Start Time: 2020-11-10T09:03:06.605+0000

* Duration: 4 min 2 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-168/detail/PR-168/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  
  
  
  
  * **Name**: `Check` (took 0 min 30 sec) . View log details on [here](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/elastic-package/pipelines/PR-168/runs/1/steps/114/log/?start=0)
    * Description: `make check`
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-11-10T09:06:38.476Z] go: downloading github.com/mattn/go-colorable v0.1.7
[2020-11-10T09:06:38.476Z] go: downloading golang.org/x/net v0.0.0-20201021035429-f5854403a974
[2020-11-10T09:06:38.476Z] go: downloading google.golang.org/appengine v1.6.6
[2020-11-10T09:06:38.476Z] go: downloading github.com/Masterminds/semver v1.5.0
[2020-11-10T09:06:38.476Z] go: downloading github.com/mitchellh/go-homedir v1.1.0
[2020-11-10T09:06:38.476Z] go: downloading github.com/elastic/package-spec/code/go v0.0.0-20201105090412-15374c6322e0
[2020-11-10T09:06:38.476Z] go: downloading github.com/spf13/pflag v1.0.5
[2020-11-10T09:06:38.476Z] go: downloading github.com/emirpasic/gods v1.12.0
[2020-11-10T09:06:38.476Z] go: downloading golang.org/x/crypto v0.0.0-20200820211705-5c72a883971a
[2020-11-10T09:06:38.476Z] go: downloading github.com/jbenet/go-context v0.0.0-20150711004518-d14ea06fba99
[2020-11-10T09:06:38.744Z] go: downloading github.com/sergi/go-diff v1.1.0
[2020-11-10T09:06:38.744Z] go: downloading golang.org/x/sys v0.0.0-20200930185726-fdedc70b468f
[2020-11-10T09:06:38.744Z] go: downloading github.com/google/go-querystring v1.0.0
[2020-11-10T09:06:38.744Z] go: downloading github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
[2020-11-10T09:06:38.744Z] go: downloading github.com/go-openapi/strfmt v0.19.6
[2020-11-10T09:06:38.744Z] go: downloading go.mongodb.org/mongo-driver v1.0.3
[2020-11-10T09:06:39.004Z] go: downloading github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
[2020-11-10T09:06:39.004Z] go: downloading github.com/mitchellh/mapstructure v1.1.2
[2020-11-10T09:06:39.004Z] go: downloading github.com/imdario/mergo v0.3.11
[2020-11-10T09:06:39.004Z] go: downloading github.com/Masterminds/semver/v3 v3.1.0
[2020-11-10T09:06:39.004Z] go: downloading github.com/mattn/go-runewidth v0.0.9
[2020-11-10T09:06:39.264Z] go: downloading github.com/xeipuuv/gojsonschema v1.2.0
[2020-11-10T09:06:39.264Z] go: downloading github.com/rakyll/statik v0.1.7
[2020-11-10T09:06:39.264Z] go: downloading github.com/go-openapi/errors v0.19.2
[2020-11-10T09:06:39.264Z] go: downloading github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415
[2020-11-10T09:06:39.264Z] go: downloading github.com/go-git/gcfg v1.5.0
[2020-11-10T09:06:39.264Z] go: downloading github.com/xeipuuv/gojsonpointer v0.0.0-20190905194746-02993c407bfb
[2020-11-10T09:06:39.264Z] go: downloading github.com/xanzy/ssh-agent v0.2.1
[2020-11-10T09:06:39.264Z] go: downloading github.com/creasty/defaults v1.5.1
[2020-11-10T09:06:39.264Z] go: downloading github.com/golang/protobuf v1.4.2
[2020-11-10T09:06:39.264Z] go: downloading gopkg.in/warnings.v0 v0.1.2
[2020-11-10T09:06:39.264Z] go: downloading github.com/kevinburke/ssh_config v0.0.0-20190725054713-01f96b0aa0cd
[2020-11-10T09:06:39.264Z] go: downloading google.golang.org/protobuf v1.25.0
[2020-11-10T09:06:39.524Z] go: downloading github.com/go-stack/stack v1.8.0
[2020-11-10T09:07:01.509Z] go get -u golang.org/x/tools/cmd/goimports
[2020-11-10T09:07:01.509Z] go: downloading golang.org/x/tools v0.0.0-20201110030525-169ad6d6ecb2
[2020-11-10T09:07:01.509Z] go: found golang.org/x/tools/cmd/goimports in golang.org/x/tools v0.0.0-20201110030525-169ad6d6ecb2
[2020-11-10T09:07:01.509Z] go: downloading golang.org/x/tools v0.0.0-20200825202427-b303f430e36d
[2020-11-10T09:07:01.509Z] go: downloading golang.org/x/mod v0.3.0
[2020-11-10T09:07:01.509Z] go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
[2020-11-10T09:07:01.509Z] go: golang.org/x/tools upgrade => v0.0.0-20201110030525-169ad6d6ecb2
[2020-11-10T09:07:01.509Z] go: golang.org/x/mod upgrade => v0.3.0
[2020-11-10T09:07:01.509Z] go: golang.org/x/xerrors upgrade => v0.0.0-20200804184101-5ec99f83aff1
[2020-11-10T09:07:01.769Z] goimports -local github.com/elastic/elastic-package/ -w .
[2020-11-10T09:07:01.769Z] go get -u golang.org/x/lint/golint
[2020-11-10T09:07:02.030Z] go: downloading golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-11-10T09:07:02.030Z] go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-11-10T09:07:02.030Z] go: golang.org/x/tools upgrade => v0.0.0-20201110030525-169ad6d6ecb2
[2020-11-10T09:07:03.417Z] go list ./... | xargs -n 1 golint -set_exit_status
[2020-11-10T09:07:06.717Z] /var/lib/jenkins/workspace/t-manager_elastic-package_PR-168/src/github.com/elastic/elastic-package/internal/testrunner/common.go:5:6: exported type StackSettings should have comment or be unexported
[2020-11-10T09:07:06.717Z] Found 1 lint suggestions; failing.
[2020-11-10T09:07:07.285Z] Makefile:15: recipe for target 'lint' failed
[2020-11-10T09:07:07.285Z] make: *** [lint] Error 123
[2020-11-10T09:07:07.361Z] Post stage
[2020-11-10T09:07:07.378Z] Running in /var/lib/jenkins/workspace/t-manager_elastic-package_PR-168/src/github.com/elastic/elastic-package
[2020-11-10T09:07:07.409Z] Archiving artifacts
[2020-11-10T09:07:07.431Z] Recording test results
[2020-11-10T09:07:07.499Z] No test report files were found. Configuration error?
[2020-11-10T09:07:07.537Z] Error when executing always post condition:
[2020-11-10T09:07:07.538Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from beats-ci-immutable-ubuntu-1804-1604998992347834357.c.elastic-ci-prod.internal/10.224.0.55:54128
[2020-11-10T09:07:07.538Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
[2020-11-10T09:07:07.538Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
[2020-11-10T09:07:07.538Z] 		at hudson.remoting.Channel.call(Channel.java:1001)
[2020-11-10T09:07:07.538Z] 		at hudson.FilePath.act(FilePath.java:1070)
[2020-11-10T09:07:07.538Z] 		at hudson.FilePath.act(FilePath.java:1059)
[2020-11-10T09:07:07.538Z] 		at hudson.tasks.junit.JUnitParser.parseResult(JUnitParser.java:111)
[2020-11-10T09:07:07.538Z] 		at hudson.tasks.junit.JUnitResultArchiver.parse(JUnitResultArchiver.java:137)
[2020-11-10T09:07:07.538Z] 		at hudson.tasks.junit.JUnitResultArchiver.parseAndAttach(JUnitResultArchiver.java:167)
[2020-11-10T09:07:07.538Z] 		at hudson.tasks.junit.pipeline.JUnitResultsStepExecution.run(JUnitResultsStepExecution.java:52)
[2020-11-10T09:07:07.538Z] 		at hudson.tasks.junit.pipeline.JUnitResultsStepExecution.run(JUnitResultsStepExecution.java:25)
[2020-11-10T09:07:07.538Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2020-11-10T09:07:07.538Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2020-11-10T09:07:07.538Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-11-10T09:07:07.538Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-11-10T09:07:07.538Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-11-10T09:07:07.538Z] hudson.AbortException: No test report files were found. Configuration error?
[2020-11-10T09:07:07.538Z] 	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:151)
[2020-11-10T09:07:07.538Z] 	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:115)
[2020-11-10T09:07:07.538Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3112)
[2020-11-10T09:07:07.538Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
[2020-11-10T09:07:07.538Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2020-11-10T09:07:07.538Z] 	at hudson.remoting.Request$2.run(Request.java:369)
[2020-11-10T09:07:07.538Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
[2020-11-10T09:07:07.538Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-11-10T09:07:07.538Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-11-10T09:07:07.538Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-11-10T09:07:07.538Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:93)
[2020-11-10T09:07:07.538Z] 	at java.lang.Thread.run(Thread.java:748)
[2020-11-10T09:07:07.538Z] 
[2020-11-10T09:07:07.916Z] Running on Jenkins in /var/lib/jenkins/workspace/t-manager_elastic-package_PR-168
[2020-11-10T09:07:08.123Z] [INFO] getVaultSecret: Getting secrets
[2020-11-10T09:07:08.246Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-11-10T09:07:09.205Z] + chmod 755 generate-build-data.sh
[2020-11-10T09:07:09.205Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-168/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-168/runs/1 FAILURE 242298
[2020-11-10T09:07:09.205Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-168/runs/1/steps/?limit=10000 -o steps-info.json
[2020-11-10T09:07:10.117Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-168/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-11-10T09:07:10.367Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2020-11-10T09:07:11.711Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2020-11-10T09:07:13.573Z] Retry 3/3 exited 22, no more retries left.
[2020-11-10T09:07:13.573Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-168/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented:
@mtojek This PR is ready for review, when you get a chance.

Before this PR we had 2 test runners, both of which operated at the data stream level and needed test configuration. This PR introduces a test runner that operates at the package level and also doesn't need test configuration. As a result, the code for the test runner framework is now a bit of a mess IMO. I would like to clean this up but in a follow up PR. WDYT?

mtojek commented on internal/kibana/ingestmanager/client_packages.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// InstallPackage installs the given package in Fleet.
+func (c *Client) InstallPackage(pkg packages.PackageManifest) ([]packages.Asset, error) {
```
We may need to unify our approaches to Kibana "subclients" as it looks like we duplicate some code. With this approach: https://github.com/elastic/elastic-package/blob/master/internal/kibana/client.go I encapsulated `dashboards` API in a separate `.go` file which shared a Kibana client. Do you have any preferences here?

The reason why I insist on the other approach is that `dashboards` API is not included within the `fleet` API (btw. we should rename it to fleet) and at some point we need to unify approaches.

Let's discuss the final design, so that this client is added in the right place in the source tree and we can adjust the rest of code.

ycombinator replied:
Agreed. In the beginning all we had was the ingestmanager (now fleet) client. But now that we have a more general kibana client, we should unify them. I took at stab at this in #210.

mtojek commented on internal/kibana/ingestmanager/client_packages.go:
```diff
@@ -0,0 +1,46 @@
+
+// RemovePackage removes the given package from Fleet.
+func (c *Client) RemovePackage(pkg packages.PackageManifest) ([]packages.Asset, error) {
+	return managePackage(pkg, "remove", c.delete)
```
I have a gut feeling that clients in `ingestmanager` are too much complex. Not sure if we need as many layers to process a single request. Maybe not depending on common abstraction would be simpler?

Please take a look at https://github.com/elastic/elastic-package/blob/master/internal/kibana/saved_objects.go . In this case I need to implement paging and the exact code responsible for calls to Kibana is wrapped up in ca. 20 LOCs: https://github.com/elastic/elastic-package/blob/master/internal/kibana/saved_objects.go#L95-L118

I'm happy to discuss it!

mtojek commented on internal/packages/assets.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	assets = append(assets, a...)
+
+	a, err = loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaMap)
```
Thinking loud: shouldn't elastic-package discover available types from the package directory? In case we add a new type the command will still operate (open-closed principle).

mtojek commented on internal/packages/assets.go:
```diff
@@ -0,0 +1,160 @@
+		return nil, errors.Wrap(err, "could not read data stream manifest file paths")
+	}
+
+	assets := make([]Asset, 0)
```
`var assets []Asset`

ycombinator replied:
Changed in f027e01.

mtojek commented on internal/packages/assets.go:
```diff
@@ -0,0 +1,160 @@
+		return nil, errors.Wrapf(err, "could not read %s files", assetType)
+	}
+
+	assets := make([]Asset, 0)
```
`var assets []Asset`

ycombinator replied:
Changed in f027e01.

mtojek commented on internal/testrunner/common.go:
```diff
@@ -0,0 +1,40 @@
+
+// GetStackSettingsFromEnv loads Elastic stack connnection settings from
+// environment variables and returns them.
+func GetStackSettingsFromEnv() StackSettings {
```
I think this is a good moment to put it in the `stack` package and expose it as `stack.CurrentSettings()`.

ycombinator replied:
Done in 23fb729.

mtojek commented on internal/testrunner/runners/asset/runner.go:
```diff
@@ -0,0 +1,167 @@
+
+// CanRunPerDataStream returns whether this test runner can run on individual
+// data streams within the package.
+func (r runner) CanRunPerDataStream() bool {
```
Let's keep it consistent with README: https://github.com/elastic/elastic-package#elastic-package-test . A context can be global, package or data stream.

ycombinator replied:
I'm not sure I follow what to do here. Would you mind elaborating?

In my mind, `elastic-package test` is a command with `package` context. It just allows for extra granularity, if the user desires, via the `--data-stream` flag.

mtojek replied:
I'm sorry but I can't remind myself what I was thinking about... Let's just ignore this comment.

mtojek commented on internal/testrunner/runners/asset/runner.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	// Install package
+	im, err := ingestmanager.NewClient(r.stackSettings.Kibana.Host, r.stackSettings.Elasticsearch.Username, r.stackSettings.Elasticsearch.Password)
```
https://github.com/elastic/elastic-package/issues/163

mtojek commented on test/packages/apache/manifest.yml:
```diff
@@ -1,7 +1,7 @@
 name: apache
 title: Apache
-version: 0.1.4
+version: 0.2.7
```
Is this change on purpose?

ycombinator replied:
Yes, I had to make it because the package registry image loaded up by the tests installs Apache version 0.2.6. So, in order for the asset test runner to be able to install the local apache package (i.e. the one being tested), I had to ensure that the local apache package's version was greater than the one in the registry. Otherwise the Install Package API would not allow the local package to be loaded up.

I admit that this is not a very maintainable change. Every time the package registry image is refreshed with a newer version of `apache`, we will need to bump up the version here as well. Maybe we should just set it here to something like `999.999.999`?

mtojek replied:
Oh, I see :) In this case you can set it to sth weird like 999. (at least the major). Nice catch!

mtojek commented on scripts/test-check-packages.sh:
```diff
@@ -10,6 +10,13 @@ cleanup() {
+for d in test/packages/*/; do
+  (
+    cd $d
+    elastic-package build -v
```
Could you please double-check if the `check` command doesn't include `build` step?

mtojek commented on internal/testrunner/testrunner.go:
```diff
@@ -141,26 +148,34 @@ func FindTestFolders(packageRootPath string, testType TestType, dataStreams []st
+}
+
+// NoConfigTestRunners returns test runners that don't need configuration.
+func NoConfigTestRunners() []TestRunner {
```
I admit I got lost here, why some test runners are defined differently (as no-config runners). I assume we can iterate on this in follow-ups.

ycombinator replied:
Yes, this was something I was experimenting with early on in this PR but is no longer necessary. I've removed it in 0e527da. Thanks for flagging it!

mtojek reviewed (commented):
@ycombinator please let me know whether I can take another look at this PR.

ycombinator commented:
> please let me know whether I can take another look at this PR.

Yeah, I'm still iterating on it. I'll re-request your review when it's ready for you to take a look. Thanks!

ycombinator commented on internal/packages/assets.go:
```diff
@@ -0,0 +1,156 @@
+
+	assets, err := loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaDashboard)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load kibana dashboard assets")
```
Done in 6bda6ef.

ycombinator commented:
I see that CI is failing. I will look into this tomorrow.

mtojek reviewed (approved):
The implementation looks good to me.

Let's move forward with this PR. Did you take a chance to run it against real integrations?

nit: just rebase the PR against master branch

ycombinator commented:
Seeing this error now:

```
$ elastic-package test -v --report-format xUnit --report-output file --defer-cleanup 1s
elastic-package has been installed.
2021/01/14 15:49:34 DEBUG Enable verbose logging
Run test suite for the package
Run asset tests for the package
2021/01/14 15:49:34 DEBUG installing package...
2021/01/14 15:49:34 DEBUG POST http://127.0.0.1:5601/api/fleet/epm/packages/apache-999.999.999
2021/01/14 15:49:34 DEBUG
2021/01/14 15:49:43 DEBUG waiting for 1s before tearing down...
Run pipeline tests for the package
Error: error parsing --data-streams flag: flag accessed but not defined: data-streams
```

Will look into it tomorrow.

ycombinator commented:
Hi @mtojek, I made a few small changes since your last review to make CI pass. Could you please take another look when you get a chance? Thanks!

mtojek commented on internal/packages/assets.go:
```diff
@@ -0,0 +1,171 @@
+	AssetTypeKibanaSavedSearch   AssetType = "search"
+	AssetTypeKibanaVisualization AssetType = "visualization"
+	AssetTypeKibanaDashboard     AssetType = "dashboard"
+	AssetTypeKibanaMap           AssetType = "map"
```
+ Lens

ycombinator replied:
Addressed in 133e50ee0364bb76f289086a0ca4aa16b92c6967.

mtojek commented on internal/testrunner/runners/system/runner.go:
```diff
@@ -51,17 +51,6 @@ type runner struct {
 	wipeDataStreamHandler   func() error
 }
 
-type stackSettings struct {
```
Is the `stackSettings` (old and new) structure used anywhere? It looks like a leftover after refactoring. All stack settings like host, login, password don't need to be exposed beyond ES/Kibana clients.

ycombinator replied:
Good catch. I've removed all stack settings code in cef5f22e598073727598a02d7324ffe5484f4f1c. So now the ES/Kibana clients directly call `os.Getenv` in their constructors to get the stack settings.

mtojek commented on internal/packages/packages.go:
```diff
@@ -200,6 +205,15 @@ func ReadDataStreamManifest(path string) (*DataStreamManifest, error) {
 
+// GetPipelineNameOrDefault returns the name of the data stream's pipeline, if one is explicitly defined in the
+// data stream manifest. If not, the default pipeline name is returned.
+func (dsm *DataStreamManifest) GetPipelineNameOrDefault() string {
```
👍 for refactoring

mtojek commented on internal/packages/packages.go:
```diff
@@ -223,5 +237,5 @@ func isDataStreamManifest(path string) (bool, error) {
 		return false, errors.Wrapf(err, "reading package manifest failed (path: %s)", path)
 	}
-	return m.Title != "" && (m.Type == "logs" || m.Type == "metrics"), nil
+	return m.Title != "" && (m.Type == dataStreamTypeLogs || m.Type == dataStreamTypeMetrics), nil
```
nit: Actually this function isn't perfect. If there is another dataset type, it will simply fail (OCP broken). I wonder if there is a different way to recognize the DataStream. If you find any, feel free to adjust the implementation.

ycombinator replied:
Yeah, it's not perfect but I think this check here might have some good side-effects. In the asset test runner we check if we are dealing with a `logs` data stream and, if so, check that ingest pipeline assets have also been loaded. Similarly, whenever a new data stream type is added in the future we might need to have special handling for it in `elastic-package`. So being strict here about which data stream types `elastic-package` knows how to handle may be a good thing. 

BTW, I was checking the data stream manifest spec to see if we have defined the data stream `type` field as an `enum` and we have. However, while doing this I noticed that there is a data stream type we have not considered in `elastic-package`: [`traces`](https://github.com/elastic/package-spec/blob/4e314c0a650da34a8f2d6f104dd659664b5377f8/versions/1/data_stream/manifest.spec.yml#L117-L120). It was added for APM in https://github.com/elastic/package-spec/pull/78. I'm guessing the team is not using `elastic-package` yet so they are not running into any issues with this type not being supported. However, I've created https://github.com/elastic/elastic-package/issues/223 to track adding support for this type in `elastic-package`.

mtojek reviewed (approved):
Thank you for the heads up! I left some minor comments, but wouldn't like to block your PR from merging. These ones can be addressed in follow up PRs (up to you).

I'm thrilled to see this running in the Integrations. Nice job!

ycombinator commented:
@mtojek I've addressed all your feedback. Please re-review when you get a chance. If the PR looks good to you, please merge it on my behalf as I'll be away from work for the next week. Otherwise I will address any further feedback when I'm back. Thanks!
